mlflow/pyspark/ml/__init__.py

```diff
@@ -0,0 +1,285 @@
+
+    param_map = {
+        param.name: instance.getOrDefault(param)
+        for param in instance.params
+        if instance.isDefined(param)
+    }
+    # TODO:
+    #  handle special case: CrossValidator.estimatorParamMaps
+
+    expanded_param_map = {}
+    for k, v in param_map.items():
+        if isinstance(v, Params):
+            # handle the case param value type inherits `pyspark.ml.param.Params`
+            # e.g. param like `OneVsRest.classifier`/`CrossValidator.estimator`
+            expanded_param_map[k] = v.uid
+            internal_param_map = _get_instance_param_map(v)
+            for ik, iv in internal_param_map.items():
+                expanded_param_map[f"{v.uid}.{ik}"] = iv
+        else:
+            expanded_param_map[k] = v
+
+    return expanded_param_map
+
+
+# NOTE: The current implementation doesn't guarantee thread-safety, but that's okay for now because:
+# 1. We don't currently have any use cases for allow_children=True.
+# 2. The list append & pop operations are thread-safe, so we will always clear the session stack
+#    once all _SparkTrainingSessions exit.
+class _SparkTrainingSession(object):
+    _session_stack = []
+
+    def __init__(self, clazz, allow_children=True):
+        self.allow_children = allow_children
+        self.clazz = clazz
+        self._parent = None
+
+    def __enter__(self):
+        if len(_SparkTrainingSession._session_stack) > 0:
+            self._parent = _SparkTrainingSession._session_stack[-1]
+            self.allow_children = (
+                _SparkTrainingSession._session_stack[-1].allow_children and self.allow_children
+            )
+        _SparkTrainingSession._session_stack.append(self)
+        return self
+
+    def __exit__(self, tp, val, traceback):
+        _SparkTrainingSession._session_stack.pop()
+
+    def should_log(self):
+        """
+        Returns True when at least one of the following conditions satisfies:
+
+        1. This session is the root session.
+        2. The parent session allows autologging and its class differs from this session's class.
+        """
+        return (self._parent is None) or (
+            self._parent.allow_children and self._parent.clazz != self.clazz
+        )
+
+
+def _get_warning_msg_for_fit_call_with_a_list_of_params(estimator):
+    return (
+        "Skip pyspark ML autologging when calling "
+        + f"{_get_fully_qualified_class_name(estimator)}.fit with a list of params,"
+        + "if you want to autolog for this case, you convert code to call `fit` with "
+        + "each single param map."
+    )
+
+
+@experimental
+@autologging_integration(FLAVOR_NAME)
+def autolog(
+    log_models=True,
+    disable=False,
+    exclusive=False,
+    disable_for_unsupported_versions=False,
+    silent=False,
+):  # pylint: disable=unused-argument
+    """
+    Enables (or disables) and configures autologging for pyspark ml estimators.
+    This method is not threadsafe and assumes a
```

Comment:
I don't think `mlflow.pyspark.ml.autolog()` assumes a SparkSession already exists, am I missing something?

Reply:
It requires SparkSession because it read the allowlist conf from sparkSession.conf.

Reply:
Can we do something like: If a spark session exists, read `sparkSesssion.conf`. If it doesn't, use the defaul allowlist ?

Reply:
> Can we do something like: If a spark session exists, read sparkSesssion.conf, otherwise use the defaul allowlist ?
I don't like it, it make the checking logic complex and may confuse user.

Because when we train spark model, we still require the spark session. if user do:

````
mlflow.pyspark.ml.autolog()
# create spark session..., if user set config here. then user may be confused why it does not work.
...
# train model
...
# log model ...
```

Reply:
Can we make the allowlist path configurable via an argument of `autolog`?

Reply:
However, auto-autolog calls `pyspark.ml.autolog` without `allowlist_path`. This makes it impossible for users to use their custom allowlist.

Reply:
I change code to adopt the solution: https://github.com/mlflow/mlflow/pull/4228#discussion_r613783280